src/librustc/util/profiling.rs

```diff
@@ -35,409 +33,136 @@ pub enum ProfilerEvent {
     QueryBlockedEnd { query_name: &'static str, category: ProfileCategory, time: u64 },
 }
 
-impl ProfilerEvent {
-    fn timestamp(&self) -> u64 {
-        use self::ProfilerEvent::*;
-
-        match self {
-            QueryStart { time, .. } |
-            QueryEnd { time, .. } |
-            GenericActivityStart { time, .. } |
-            GenericActivityEnd { time, .. } |
-            QueryCacheHit { time, .. } |
-            QueryCount { time, .. } |
-            IncrementalLoadResultStart { time, .. } |
-            IncrementalLoadResultEnd { time, .. } |
-            QueryBlockedStart { time, .. } |
-            QueryBlockedEnd { time, .. } => *time
-        }
-    }
+#[repr(u32)]
+#[derive(Clone, Copy, Debug)]
+enum EventKind {
+    Query = (u32::MAX / 2) - 1,
+    GenericActivity = (u32::MAX / 2) - 2,
+    IncrementalLoadResult = (u32::MAX / 2) - 3,
+    QueryBlocked = (u32::MAX / 2) - 4,
+    QueryCacheHit = (u32::MAX / 2) - 5,
 }
 
 fn thread_id_to_u64(tid: ThreadId) -> u64 {
     unsafe { mem::transmute::<ThreadId, u64>(tid) }
 }
 
 pub struct SelfProfiler {
-    events: FxHashMap<ThreadId, Vec<ProfilerEvent>>,
-    start_time: SystemTime,
-    start_instant: Instant,
+    profiler: Profiler<FileSerializationSink>,
 }
 
 impl SelfProfiler {
     pub fn new() -> SelfProfiler {
-        let profiler = SelfProfiler {
-            events: Default::default(),
-            start_time: SystemTime::now(),
-            start_instant: Instant::now(),
+        let filename = format!("pid-{}.rustc_profile", process::id());
+        let path = std::path::Path::new(&filename);
+        let profiler = Profiler::new(path);
+
+        let event_kinds = &[
+            (EventKind::Query, "Query"),
+            (EventKind::GenericActivity, "GenericActivity"),
+            (EventKind::IncrementalLoadResult, "IncrementalLoadResult"),
+            (EventKind::QueryBlocked, "QueryBlocked"),
+            (EventKind::QueryCacheHit, "QueryCacheHit"),
+        ];
+
+        for (kind, s) in event_kinds {
+            let string_id = StringId::reserved(*kind as u32);
+            profiler.alloc_string_with_reserved_id(string_id, *s);
+        }
```

Comment:
I think I would get rid of the `EventKind` enum and just use a regular `alloc_string` here, storing the `StringId`s in fields of the `SelfProfiler` struct. That should have about the same performance and doesn't rely on magic numbers.